### PR TITLE
Disable default allowing the testing of connections in UI, API and CLI

### PR DESCRIPTION
**Pull request by pankajkoti**
With this PR, the test connection functionality will be disabled
by default across Airflow UI, API and CLI. The availability of the 
functionality can be controlled by the `test_connection` flag in 
the `core` section of the Airflow configuration (`airflow.cfg`). 
It can also be controlled by the environment variable 
`AIRFLOW__CORE__TEST_CONNECTION`.

The following values are accepted for this config param:
1. `Disabled`: Disables the test connection functionality and 
disables(greys out) the Test Connection button in the UI.
This is also the default value set in the Airflow configuration.
2. `Enabled`: Enables the test connection functionality and 
activates the Test Connection button in the UI.
3. `Hidden`: Disables the test connection functionality and 
hides the Test Connection button in UI.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/config_templates/config.yml
Comment (o-nikolas) [inline airflow/config_templates/config.yml]: Wording suggestion, feel free to take it or not :) 
```suggestion
        Enables or disables the ability to test connections from UI. By default, it is disabled.
```

File: airflow/config_templates/config.yml
Comment (o-nikolas) [inline airflow/config_templates/config.yml]: Another suggested wording, feel free to use or not (if used here, be sure to also copy it below in default config)
```suggestion
        Before setting this to True, make sure that you review the users who are able
        to add/edit connections in the UI and ensure they are trusted.
        Connection testing can can be done maliciously leading to undesired and insecure outcomes.
```

File: airflow/www/static/js/connection_form.js
Comment (o-nikolas) [inline airflow/www/static/js/connection_form.js]: Another suggestion
```suggestion
      .attr("title", "Testing connections is disabled in Airflow configuration.");
```

**Review by o-nikolas — COMMENTED** [review]
Nice!

File: airflow/www/static/js/connection_form.js
Comment (jedcunningham) [reply airflow/www/static/js/connection_form.js]: I wonder if we should just get rid of the button altogether. I think that'd be my vote.

**Review by jedcunningham — COMMENTED** [review]
We also need to disable endpoints that serve this functionality! It can't be a UI only thing.

File: airflow/www/static/js/connection_form.js
Comment (potiuk) [reply airflow/www/static/js/connection_form.js]: I think It's kinda good to see it so that it is discoverable that you can do it and explain that you need to talk to your admin to enable it (as a tooltip). 

This is especially important, I think, for existing users. If suddenly the button disappears, nobody (especially users who could use the "Test connection" before) will read the release notes or documentation, we will be opening for floodgate of issues "The `test connection` button disappeared - please fix it".

Comment (potiuk) [reply airflow/www/static/js/connection_form.js]: My suggestion will also be to add. "Talk to your deployment manager to enable it." or similar. I think it is time to start introducing the different kinds of users for our security model and once we explain the different kinds of users, use it consistently as messaging. 

I think having clear responsibilities "who is responsible for what" is also part of the user "education" we have to do a bit.  The average user will not know that they need to talk to the "deplomen manager" I think (if they are on a corporate installation or managed service). Which config? How to enable it? Where to find it ? 

I think there are often just "users" who who do not know (and should not even need to know actually) what Airflow configuration is. But if we tell them "talk to the guys that manage your Airflow", they will know what to do  (rather than trying to look on their own or create issues in airflow repo "I see this messsage, what should i do" (happens).

Comment (jedcunningham) [reply airflow/www/static/js/connection_form.js]: Good point, but playing devils advocate :) I wonder if we need a non-boolean value for this then? What about the admin who chooses not to allow their users to test connections but constantly gets asked about it? Allowing admins to hide it could save them the same heartache.

Comment (pankajkoti) [reply airflow/www/static/js/connection_form.js]: okay, so maybe we can give another config param to allow removing the button altogether? 
`remove_test_connection_button` set by default to `False` meaning this button will appear by default.

And `enable_test_connection=False` would control enabling/disabling the functionality

For evaluation:
remove_test_connection_button = Priority 0
enable_test_connection = Priority 1

0 has higher priority than 1. Would that make sense?

Comment (pankajkoti) [reply airflow/www/static/js/connection_form.js]: or `test_connection` = enable/disable/remove. disable being the default?

Comment (jedcunningham) [reply airflow/www/static/js/connection_form.js]: I'd strongly prefer a single config for this (we have enough config options already).

Comment (potiuk) [reply airflow/www/static/js/connection_form.js]: Yeah. Very much agree with @jedcunningham  :). And agree that being able to disable it completely is a good idea from Deployment manager's perspective. I am torn between the defaults there though. Should new airflow has "remove" by defult or "disable". If only we could see if we are upgrading from older version or installing airflow from the scratch, that would be cool to have `remove` by default and `disable` when you migrate.

Comment (o-nikolas) [reply airflow/www/static/js/connection_form.js]: To play the other Devil's advocate, I think complicating this config beyond a boolean might be over-engineering this. I think having the button there, but greyed out, is useful for discoverability, and I don't think folks will really harass the deployment manager much in reality because of it.

Comment (potiuk) [reply airflow/www/static/js/connection_form.js]: 👍 🤔 👎 

I am ok with either approach as long as we have it off by default :)

File: tests/www/views/test_views.py
Comment (pankajkoti) [inline tests/www/views/test_views.py]:
```suggestion
```

File: airflow/www/static/js/connection_form.js
Comment (pankajkoti) [reply airflow/www/static/js/connection_form.js]: A bit confused about what could be the best approach here. But observed that the UI treats the values as a string anyway to do a comparison of the value. 

So, based on the conversation above and considering @jedcunningham's and @potiuk's suggestions, I thought to provide a single config param `test_connection` in the `core` section of Airflow configuration as this param needs to be used across Airflow UI, API and CLI. The supported values I have implemented are `Disabled`, `Enabled` and `Hidden` with `Disabled` being the default. @o-nikolas hope you are okay with this approach 🤔

File: airflow/config_templates/config.yml
Comment (potiuk) [inline airflow/config_templates/config.yml]:
```suggestion
      version_added: 2.7.0
```

I think (but I would love to hear what others think) - it's a bit too disruptive to release it in PATCHLEVEL release. I see this one as a long-term decision that should impact whether test connection is enabled or not, and there is no immediate need to add it in patchlevel version IMHO.

File: airflow/config_templates/config.yml
Comment (potiuk) [inline airflow/config_templates/config.yml]: Seems they were added by some auto-formatting. Can we remove those empty lines back?

**Review by potiuk — COMMENTED** [review]
Few small nits, but LGTM

File: airflow/config_templates/config.yml
Comment (pankajkoti) [reply airflow/config_templates/config.yml]: Actually, I added them manually 🙃 . I observed in some places we had empty new lines between these config sections, in some places no empty new lines to separate the sections and in 1-2 occurrences multiple empty new lines between sections. How would we like to keep it?
Remove the empty new line separators between all sections?

File: airflow/config_templates/config.yml
Comment (pankajkoti) [reply airflow/config_templates/config.yml]: That makes sense, will keep it for 2.7.0 unless someone says we need it for 2.6.3.  Will make this change in my next commit.

Edit: The commit with the change to 2.7.0 has now been pushed.

File: airflow/config_templates/config.yml
Comment (pankajkoti) [reply airflow/config_templates/config.yml]: Removed empty new lines from everywhere else to keep it consistent.

**Review by potiuk — APPROVED** [review]

File: airflow/www/static/js/connection_form.js
Comment (o-nikolas) [reply airflow/www/static/js/connection_form.js]: Yupp, that will be perfectly fine :)

**Review by o-nikolas — APPROVED** [review]

**Comment by potiuk** [comment]
We also need a significant newsfragment for that one @pankajkoti

**Comment by pankajkoti** [comment]
> We also need a significant newsfragment for that one @pankajkoti

Thank you. Naive to add this one. Added one by following the guideline [here](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#step-4-prepare-pr) and taking reference of [this PR](https://github.com/apache/airflow/pull/31277/files#diff-527c5b0190068f64dc65bdf2b25137ca746178bf33f811aeb5bd4928dfe1d97e) which says it can be like a git commit message for significant types.

**Comment by pankajkoti** [comment]
The tests are failing because of an unrelated error
```
=========================== short test summary info ============================
FAILED tests/providers/amazon/aws/triggers/test_athena.py::TestAthenaTrigger::test_run_success - TypeError: async_wait() got an unexpected keyword argument 'max_attempts'
=========== 1 failed, 1890 passed, 19 warnings in 898.71s (0:14:58) ============
```

Looks like the PR which added this change is reverted now. I will rebase now.

**Comment by potiuk** [comment]
One more small comment. I've been thiknig about it.. Currently we have this but only in comments. 

```
Before setting this to Enabled, make sure that you review the users who are able to add/edit
        connections and ensure they are trusted. Connection testing can be done maliciously leading to
        undesired and insecure outcomes.
```

Should we make also add some more information (link to the right part of the security model documentation once we merge https://github.com/apache/airflow/pull/32098 should be enough ) - both in the newsfragment (effectively release notes) and in the place in documentation where we explain test connection, to explain why it is dangerous to enable test connection?

I think we should be very explicit about it if we want to make our model influence decisions of our users.

**Comment by pankajkoti** [comment]
yes makes sense. Which part of the PR would we like to link it to? Some section in this file `docs/apache-airflow/security/index.rst `? 

Will wait for the PR to be merged and then link it here.

**Comment by potiuk** [comment]
> yes makes sense. Which part of the PR would we like to link it to? Some section in this file docs/apache-airflow/security/index.rst ?

Likely. `docs/apache-airflow/security/index.rst#capabilities_of_authenticated_ui_users` if I got it right :)

**Comment by potiuk** [comment]
Or maybe better will be to add reference and link to it. I might add it now.

**Review by potiuk — APPROVED** [review]

**Pull request merged**
